### ISSUE TITLE
[Mac] Invalid form params when multiple submit buttons are used

### ISSUE THREAD

**Issue by makmic** [unpoly/unpoly#103]
Given this example:
```html
<form method="post" action="/my/action" up-target=".post">
  <button type"submit">Save</button>
  <button type"submit" name="my-param" value="my-value">Do something else</button>
</form>
```
On most systems, submitting the form by pressing the second button sends a POST request to `/my/action` with the parameters ` {'my-param': 'my-value'}`.
On Mac devices (Safari or Firefox), however, always the value of the first submit button is part of the Params hash. Chrome on Mac works fine.

Digging in Unpoly's code, I found that the method [submittingButton(form)](https://github.com/unpoly/unpoly/blob/cf91465eb9081256a69be4439832faefd1547013/lib/assets/javascripts/unpoly/form.coffee.erb#L92) in this case contains the wrong submit button:

```javascript
  submittingButton = (form) ->
    selector = submitButtonSelector()
    focusedElement = document.activeElement
    if focusedElement && e.matches(focusedElement, selector) && form.contains(focusedElement)
      return focusedElement
    else
      # If no button is focused, we assume the first button in the form.
      return e.first(form, selector)
```

The condition for the `if` clause matches in Chrome but fails in Safari or FF on a Mac: `focusedElement` is not the targeted button, but the HTML element. The property `document.activeElement` works only reliably for `input[type="text"]` elements. The [MDN Reference](https://developer.mozilla.org/en-US/docs/Web/API/DocumentOrShadowRoot/activeElement) states:

> Which elements are focusable varies depending on the platform and the browser's current configuration. For example, on **macOS systems, elements that aren't text input elements are not typically focusable by default**.

I am not sure what would be the best solution to this issue. The browser market share of Safari is around 15%, so we might need to work around this.

**Comment by makmic** [comment]
A note for users affected by this issue: To work around it, you could temporarity add this compiler to your application, which allows form buttons to be focused on a Mac.

```javascript
// This is a workaround for an Unpoly issue on macOS (Safari and Firefox)
// On Mac OS (by design) buttons will not get focus when clicked
// so Unpoly fails to detect which button was clicked, and hence will
// not serialize it for the XHTML request.

up.compiler('form[up-target] button[name]', (element) => {
  element.addEventListener('click', () => {
    element.focus();
  });
});
```

**Comment by naturallymitchell** [comment]
How might someone use multiple buttons in a form?

**Comment by triskweline** [comment]
> How might someone use multiple buttons in a form?

For example, a form that lets you approve or deny a record might have two buttons "Approve" and "Deny". The `value` attribute of the pressed button should then be submitted together with the other form values.

**Comment by pfw** [comment]
I've been having a poke at this seeing if I could help, I've been stung by this as well, and don't have a suitable fix yet but have some more info for Chrome/Edge/Firefox/Safari on Mac.

Within a 'submit' handler:
- 'event.submitter' is set to the button clicked on Chrome/Edge/Firefox
- Safari doesn't set that but does set the button name=value in FormData(form) created within the handler
- Other browsers don't include the button in a FormData(form) created in the handler

Basically it is either set as event.submitter or included in FormData(form) but does require having access to the submit event.

Chrome 81
Edge 83
Firefox 76
Safari 13.1

**Comment by pfw** [comment]
@makmic As an aside, an alternative pattern is available through 'formaction=URL' on the button. I'd not run across this before but I think is very interesting as it enables the form to be sent to a different URL/controller when different buttons are pressed, skipping the need for "if button-x pressed..." etc. logic in a single form controller. It appears to be quite well supported at least back to ie10.

**Comment by triskweline** [comment]
This is fixed in [Unpoly 2](https://unpoly.com/changes/2.0.0), which was released today.
